### ISSUE TITLE
Yesod.Auth.Email RegisterR sends verification email if the email is already registered

### ISSUE THREAD

**Issue by sevcsik** [yesodweb/yesod#948]
I found out when I try to register an email on RegisterR which already exists, instead of displaying an error message, that the user already exists, it sends an an email confirmation request to the given email address, just like if I asked for a password reset.

After going through the sources, I found out that `registerHelper` checks if the email exists, and if it does, it sends a verify email even if it's already verified.

``` haskell
    -- [...]
        mecreds <- lift $ getEmailCreds identifier                           
            registerCreds <-                                                 
                case mecreds of                                              
                    Just (EmailCreds lid _ _ (Just key) email) -> return $ Just (lid, key, email)
                    Just (EmailCreds lid _ _ Nothing email) -> do            
                        key <- liftIO $ randomKey y                          
                        lift $ setVerifyKey lid key                          
                        return $ Just (lid, key, email)                      
                    Nothing                                                  
                        | allowUsername -> return Nothing                    
                        | otherwise -> do                                    
                            key <- liftIO $ randomKey y                      
                            lid <- lift $ addUnverified identifier key       
                            return $ Just (lid, key, identifier)

            case registerCreds of
                Nothing -> loginErrorMessageI dest (Msg.IdentifierNotFound identifier)
                Just (lid, verKey, email) -> do
                    render <- getUrlRender
                    let verUrl = render $ verify (toPathPiece lid) verKey
                    lift $ sendVerifyEmail email verKey verUrl
                    lift $ confirmationEmailSentResponse identifier                                         
    -- [...]
```

Source: https://hackage.haskell.org/package/yesod-auth-1.4.3.1/docs/src/Yesod-Auth-Email.html#registerHelper

Since `registerCreds` is a `Just` when the email exists and it has a key, it sends the verify email again to the user, whiche he/she can use as a password reset link. 

I think the first two cases of `mecreds` should be something like this:

``` haskell
Just (EmailCreds lid _ False (Just key) email) -> return $ Just (lid, key, email)
Just (EmailCreds _ _ True _ _) -> return Nothing
```

So if the email is not yet verified, it just gets the verification key sent again, but if it is verified already, it displays an error message instead of acting like a password reset, which is a bit confusing.

Also, it would be nice to put `registerHelper` under the `YesodAuthEmail` typeclass, so this behavior could be customized.

**Comment by geraldus** [comment]
good catch!

**Comment by Laquendi** [comment]
There is my suggestion with arguably nicer API than just exposing registerHelper.
I can't test it right now since I don't have a working mailer and I need to sleep.

**Comment by sevcsik** [comment]
@Laquendi: thanks, the new API looks great

**Comment by snoyberg** [comment]
Just to understand this better: you're saying that:

* When registering, you only get a verify email if the email is not verified
* When requesting a password reset, you get a verify email only if the email is already verified

Is that the goal?

Personally, I actually prefer the current behavior, as it addresses the case where someone forgot whether they have an account already. I understand that not everyone wants to have that behavior, but I actually *like* getting rid of the distinction between register and verify (I've had many experiences with that just causing confusion).

I'm in favor of some kind of a method in the typeclass to allow this behavior to be configurable.

**Comment by sevcsik** [comment]
The current behaviour does makes sense, I didn't even treat it as a bug
initially, but asked on SO if it can be changed.

It does work well, and from a technical perspective it is valid. But I
think a non-technical user would find it confusing. If I register at a
site, I except two outcomes: either my registration goes through, or I get
an error. I wouldn't expect the site to go ahead and do the password reset
process for me without asking.

So I would prefer a way to change this behaviour, so the registration user
journey can be similar to most of the sites, which the masses are already
accustomed to.

**Comment by snoyberg** [comment]
I'd say we're looking at a few specific switches in behavior; we've identified two already (current behavior and your proposal), and we'll likely have a few others in the future. Let's introduce something like this:

```haskell
data VerifyEmailMode = ... -- abstract data type, constructors not exported

alwaysSendVerify :: VerifyEmailMode -- current behavior
onlySendOnUnverified :: VerifyEmailMode -- new proposed behavior

-- Inside the typeclass
    verifyEmailMode :: a -> VerifyEmailMode
    verifyEmailMode _ = alwaysSendVerify
```

Names completely up for debate, I have no affinity to the ones I've provided.

**Comment by sevcsik** [comment]
Looks good! One more thing I noticed: probably we also need a new message
en Yesod.Auth.Messages for this case, because none of the current ones
looks applicable.

**Comment by snoyberg** [comment]
Now that things are specced out, this should be a pretty simple change for someone to make. PR welcome.

**Comment by StevenXL** [comment]
Hi folks. I wanted to contribute this as a PR, but the code is not ready and I am a bit stuck on how to proceed. I have created a [PR against my own repo's master branch](https://github.com/StevenXL/yesod/pull/1) for easier comparison. These are the questions I have right now.

First, on Line 210 of the PR, I am confused as to what version to write down. The `cabal` file says that the library is currently on `1.6.3`, but I don't know when the PR will be merged in. Should I just go with `1.6.4`?

Secondly, On Line 555, I have not yet implemented the case in which the email is already verified, and we have set the verify email mode where we don't want to resend the email. I believe what we want is a value of type `AuthHandler site TypeContent`, with a JSON response and an HTML response (wrapped in the default layout). The message would be a new message ("Email has already been verified"?), but I'd only be able to provide in Spanish and English. Would that suffice?

Finally, this wouldn't stop me from making a PR, but since we are on the subject, I'm confused as to how I can apply the `y` value on Line 540. `getYesod :: m0 (HandlerSite m0)`. `y` then has the type `HanderlSite m0`, for some type `m0`. So `HandlerSite m0`, whatever `m0` is must be an instance of `YesodAuthEmail`, yes?
